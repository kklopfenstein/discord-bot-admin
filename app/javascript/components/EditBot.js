import axios from 'axios';
import React, { useEffect, useState } from 'react';
import { useNavigate, useParams, Link } from 'react-router-dom';
import BotForm from './BotForm';
import Button from '@mui/material/Button';

export default () => {
  const [formErrors, setFormErrors] = useState();
  const [bot, setBot] = useState();
  const params = useParams();

  const navigate = useNavigate();

  const onSubmit = (data) => {
    const updateData = {...data, id: params.botId };
    axios.put(`/api/v1/bots/${updateData.id}`, { bot: updateData })
    .then(() => {
      navigate('/bots');
    })
    .catch((error) => {
      console.error(error);
      setFormErrors(
        <div>An error occurred.</div>
      )
    });
  };

  const loadBot = (id) => {
    axios.get(`/api/v1/bots/${id}`)
    .then((response) => {
      setBot(response.data);
    })
    .catch((error) => {
      console.error(error);
      setFormErrors(
        <div>The bot could not be found.</div>
      )
    });
  };


  useEffect(() => {
    loadBot(params.botId);
  }, [params.id])

  let response = <div>Loading...</div>

  if (bot) {
    response = <BotForm onSubmit={onSubmit} bot={bot} formErrors={formErrors}/>
  }

  return (
    <>
    <Link to="/bots">
      <Button variant="outline">
        Return to bots
      </Button>
    </Link>
    {response}
    </>
  );
};