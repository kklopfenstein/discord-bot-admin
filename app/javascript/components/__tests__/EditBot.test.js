import React from 'react';
import EditBot from '../EditBot';
import {fireEvent, render, screen, act} from '@testing-library/react';
import '@testing-library/jest-dom'
import renderer, { act as actRenderer } from 'react-test-renderer';
import mockAxios from 'jest-mock-axios';

jest.mock('react-router-dom', () => ({
  Link: () => {
    const MockName = "mock-name";
    return <MockName />;
  },
  useNavigate: jest.fn(),
  useParams: () => {
    return {
      botId: 1
    };
  }
}));

describe('EditBot', () => {
  afterEach(() => {
    mockAxios.reset();
  });

  it('displays form', () => {
    let component;
    
    actRenderer(() => {
      component = renderer.create(
        <EditBot />
      );
    });
    
    expect(mockAxios.get).toHaveBeenCalledWith('/api/v1/bots/1');
    let responseObj = { 
      data: {
        id: 1,
        name: "edited item",
        token: "token"
      }
    };

    actRenderer(() => {
      mockAxios.mockResponse(responseObj);
    });

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });


  it('submits the form posts to api', async () => {
    const { container } = render(<EditBot/>);

    expect(mockAxios.get).toHaveBeenCalledWith('/api/v1/bots/1');
    let responseObj = { 
      data: {
        id: 1,
        name: "edited item",
        token: "OTM532IwNzgyMDEzNDZhhhgw.Yd6C6Q.Pslx7BdYAlmCsTkJDndWHv8be3z"
      }
    };

    act(() => {
      mockAxios.mockResponse(responseObj);
    });

    const name = container.querySelector("input[name='name']");
    const token = container.querySelector("input[name='token']");
    const submitButton = container.querySelector("button");

    fireEvent.input(name, {
      target: {
        value: 'test'
      }
    });

    fireEvent.input(token, {
      target: {
        value: 'OTM532IwNzgyMDEzNDZhhhgw.Yd6C6Q.Pslx7BdYAlmCsTkJDndWHv8be3z'
      }
    });

    await act(async () => {
      fireEvent.click(submitButton);
    });

    expect(mockAxios.put).toHaveBeenCalledWith('/api/v1/bots/1', {'bot': {'id': 1, 'token': 'OTM532IwNzgyMDEzNDZhhhgw.Yd6C6Q.Pslx7BdYAlmCsTkJDndWHv8be3z', 'name': 'test'}});
  });
});