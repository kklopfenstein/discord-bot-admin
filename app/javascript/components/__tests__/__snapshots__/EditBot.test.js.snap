// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditBot displays form 1`] = `
Array [
  <mock-name />,
  <form>
    <div
      className="MuiBox-root css-1qapnz0"
    >
      <div
        className="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
      >
        <div
          className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor="mui-1"
            id="mui-1-label"
          >
            Name
          </label>
          <div
            className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              disabled={false}
              id="mui-1"
              name="name"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value="edited item"
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-173wfxe"
              >
                <span>
                  Name
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div />
      </div>
    </div>
    <div
      className="MuiBox-root css-1qapnz0"
    >
      <div
        className="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
      >
        <div
          className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor="mui-2"
            id="mui-2-label"
          >
            Token
          </label>
          <div
            className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              disabled={false}
              id="mui-2"
              name="token"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value="token"
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-173wfxe"
              >
                <span>
                  Token
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div />
      </div>
    </div>
    <div
      className="MuiBox-root css-1qapnz0"
    >
      <button
        className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Submit
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>,
]
`;
